cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginService.java

```diff
@@ -29,32 +30,41 @@
 import co.cask.cdap.proto.Id;
 import co.cask.cdap.proto.artifact.ArtifactRange;
 import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
 import com.google.common.io.Closeables;
 
 import java.io.Closeable;
 import java.io.File;
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import javax.ws.rs.Path;
 
 /**
  * To find, instantiate and invoke methods in plugin artifacts
  */
-public class PluginService implements Closeable {
+public class PluginService {
   private final ArtifactRepository artifactRepository;
   private final File stageDir;
   private final CConfiguration cConf;
 
-  private CloseableClassLoader parentClassLoader;
-  private PluginInstantiator pluginInstantiator;
+  private final LoadingCache<ArtifactDescriptor, Instantiators> instantiators;
 
   public PluginService(ArtifactRepository artifactRepository, File tmpDir, CConfiguration cConf) {
     this.artifactRepository = artifactRepository;
     this.stageDir = DirUtils.createTempDir(tmpDir);
     this.cConf = cConf;
+    this.instantiators = CacheBuilder.newBuilder()
```

Comment:
The `PluginInstantiator` will get close (and invalidating the cache) when the program execution finished, it is per container and only contains plugins that are used by the program execution. However, this class is inside app-fabric and is caching all possible plugins.